### PR TITLE
[ONNX] Fix rotary_embedding_23 implementation

### PR DESCRIPTION
**Pull request by justinchuby**
The implementation of rotary_embedding_23 when input is 3D was incorrect.

## Tested

Locally with 

```py
import onnx_ir as ir
import onnx
import torch
import os
import numpy as np

base_path = "/home/justinchu/dev/onnx/onnx/backend/test/data/node"
test_names = [
    "test_rotary_embedding",
    "test_rotary_embedding_3d_input",
    "test_rotary_embedding_interleaved",
    "test_rotary_embedding_no_position_ids",
    "test_rotary_embedding_no_position_ids_interleaved",
    "test_rotary_embedding_no_position_ids_rotary_dim",
    "test_rotary_embedding_with_interleaved_rotary_dim",
    "test_rotary_embedding_with_rotary_dim",
]
model_paths = [os.path.join(base_path, name) for name in test_names]


for path in model_paths:
    print(f"Checking {path} for issues...")

    model = onnx.load(os.path.join(path, "model.onnx"))
    input0 = ir.from_proto(
        onnx.load_tensor(os.path.join(path, "test_data_set_0", "input_0.pb"))
    ).numpy()
    input1 = ir.from_proto(
        onnx.load_tensor(os.path.join(path, "test_data_set_0", "input_1.pb"))
    ).numpy()
    input2 = ir.from_proto(
        onnx.load_tensor(os.path.join(path, "test_data_set_0", "input_2.pb"))
    ).numpy()
    if os.path.exists(os.path.join(path, "test_data_set_0", "input_3.pb")):
        input3 = ir.from_proto(
            onnx.load_tensor(os.path.join(path, "test_data_set_0", "input_3.pb"))
        ).numpy()
    else:
        input3 = None
    output0 = ir.from_proto(
        onnx.load_tensor(os.path.join(path, "test_data_set_0", "output_0.pb"))
    ).numpy()

    m = ir.from_proto(model)

    node = m.graph[-1]
    print(node)
    assert node.op_type == "RotaryEmbedding"

    interleaved = node.attributes.get_int("interleaved", 0)
    num_heads = node.attributes.get_int("num_heads", 0)
    rotary_embedding_dim = node.attributes.get_int("rotary_embedding_dim", 0)

    torch_out = torch.onnx.ops.rotary_embedding(
        torch.tensor(input0),
        torch.tensor(input1),
        torch.tensor(input2),
        position_ids=torch.tensor(input3) if input3 is not None else None,
        interleaved=bool(interleaved),
        num_heads=num_heads,
        rotary_embedding_dim=rotary_embedding_dim,
    )
    torch_out = torch_out.detach().cpu().numpy()
    np.testing.assert_allclose(torch_out, output0)
```

Fix https://github.com/pytorch/pytorch/issues/162848

cc @titaiwangms @kunal-vaishnavi

### CODE REVIEW

**Comment by justinchuby** [comment]
Cross reference
- https://github.com/onnx/onnx/blob/be3e3799fbd0acbc654a9579498ec875b129588d/onnx/defs/nn/defs.cc#L3231
- https://onnx.ai/onnx/operators/onnx__RotaryEmbedding.html#rotaryembedding-23
- https://github.com/onnx/onnx/blob/main/onnx/reference/ops/op_rotary_embedding.py

File: torch/onnx/ops/_impl.py [hunk 1]
```diff
@@ -56,18 +56,54 @@ def rotary_embedding_23(
     rotary_embedding_dim: int = 0,
 ) -> torch.Tensor:
     """RotaryEmbedding-23 https://onnx.ai/onnx/operators/onnx__RotaryEmbedding.html#rotaryembedding-23"""
+    # x has shape (batch_size, num_heads, sequence_length, head_size)
+    # or (batch_size, sequence_length, hidden_size)
+    input_shape = x.shape
+    input_rank = len(input_shape)
+    batch_size = input_shape[0]
+    sequence_length = input_shape[-2]
```
Comment (titaiwangms) [inline torch/onnx/ops/_impl.py:64]: If 4d, this is input_shape[-1]

File: torch/onnx/ops/_impl.py [hunk 2]
```diff
@@ -56,18 +56,54 @@ def rotary_embedding_23(
...
+        )
+        torch._check(
+            cos_cache.dim() == 2 and sin_cache.dim() == 2,
+            lambda: "cos_cache/sin_cache must be 2D (sequence_length, head_half) when position_ids is provided. "
+            f"Got cos_cache shape {cos_cache.shape}, sin_cache shape {sin_cache.shape}",
+        )
+    else:
+        torch._check(
+            cos_cache.dim() == 3 and sin_cache.dim() == 3,
+            lambda: "cos_cache/sin_cache must be 3D (batch, sequence_length, head_half) when position_ids is not provided",
```
Comment (titaiwangms) [inline torch/onnx/ops/_impl.py]: Should we test all possible combinations? For example, these lines provide comprehensive checks: https://github.com/microsoft/onnxruntime/blob/5d1773454d38bb6297a6a4d77c9c8cad872ac873/onnxruntime/core/providers/cpu/llm/rotary_embedding_helper.h#L82-L167

File: torch/onnx/ops/_impl.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/ops/_impl.py:64]: I though sequence_length is always [-2]? When 4d it is (batch_size, num_heads, sequence_length, head_size)

File: test/onnx/ops/test_ops.py [hunk 3]
```diff
@@ -525,6 +532,49 @@ def forward(
         )
         self.assertEqual(onnx_program.model.opset_imports[""], 23)
         self.assertEqual("RotaryEmbedding", onnx_program.model.graph.node(0).op_type)
+        if has_onnxruntime_opset_23():
+            onnx_testing.assert_onnx_program(onnx_program)
+        else:
+            # Test with reference evaluator because ORT does not support the op as of version 1.22
+            onnx_testing.assert_onnx_program(onnx_program, backend="reference")
```
Comment (xadupre) [inline test/onnx/ops/test_ops.py:539]: Maybe you could test both implementations.

Comment (justinchuby) [reply test/onnx/ops/test_ops.py:539]: Do you know which ort release has the op?

File: torch/onnx/ops/_impl.py [hunk 4]
```diff
@@ -56,18 +56,54 @@ def rotary_embedding_23(
...
+            position_ids.shape[0] == batch_size,
+            lambda: f"position_ids first dim (batch) must match x.shape[0] ({batch_size}). Received {position_ids.shape[0]}",
+        )
+        torch._check(
+            position_ids.shape[1] == sequence_length,
+            lambda: f"position_ids second dim (sequence) must match x.shape[-2] ({sequence_length}). Received {position_ids.shape[1]}",
+        )
+        torch._check(
+            cos_cache.dim() == 2 and sin_cache.dim() == 2,
+            lambda: "cos_cache/sin_cache must be 2D (sequence_length, head_half) when position_ids is provided. "
```
Comment (jambayk) [inline torch/onnx/ops/_impl.py]: first dim is `max_position_id_plus_1`

File: torch/onnx/ops/_impl.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/ops/_impl.py:64]: Oh yeah, nvm

File: torch/onnx/ops/_impl.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/ops/_impl.py]: Removed the shape reference because there are four cases.

**Review by kunal-vaishnavi — APPROVED** [review]

**Review by titaiwangms — APPROVED** [review]
